pkg/controllers/node/ipam.go

```diff
@@ -62,63 +65,268 @@ func registerPrometheusMetrics() {
 	prometheus.MustRegister(blocksGauge)
 }
 
-// syncIPAMCleanup cleans up any IPAM resources which should no longer exist based on nodes in the cluster.
-// It returns an error if it is determined that there are resources which should be cleaned up, but is unable to do so.
-// It does not return an error if it is successful, or if there is no action to take.
-func (c *NodeController) syncIPAMCleanup() error {
-	log.Info("Synchronizing IPAM data")
-	data, err := c.gatherIPAMData()
-	if err != nil {
-		return err
+func NewIPAMController(cfg config.NodeControllerConfig, c client.Interface, cs *kubernetes.Clientset) *ipamController {
+	return &ipamController{
+		client:      c,
+		clientset:   cs,
+		config:      cfg,
+		rl:          workqueue.DefaultControllerRateLimiter(),
+		syncChan:    make(chan interface{}, 1),
+		blockUpdate: make(chan model.KVPair),
+
+		allBlocks:          make(map[string]model.KVPair),
+		allocationsByBlock: make(map[string]map[string]*allocation),
+		allocationsByNode:  make(map[string]map[string]*allocation),
 	}
-	err = c.cleanup(data)
-	if err != nil {
-		return err
+
+}
+
+type ipamController struct {
+	rl        workqueue.RateLimiter
+	client    client.Interface
+	clientset *kubernetes.Clientset
+	config    config.NodeControllerConfig
+
+	syncStatus bapi.SyncStatus
+
+	// syncChan triggers processing in response to an update.
+	syncChan chan interface{}
+
+	// For block update / deletion events.
+	blockUpdate chan model.KVPair
+
+	// Raw block storage.
+	allBlocks map[string]model.KVPair
+
+	// Store allocations broken out from the raw blocks by their handle.
+	allocationsByBlock map[string]map[string]*allocation
+	allocationsByNode  map[string]map[string]*allocation
+}
+
+func (c *ipamController) Start(stop chan struct{}) {
+	go c.acceptScheduleRequests(stop)
+
+	// Trigger a start-of-day sync.
+	kick(c.syncChan)
+}
+
+func (c *ipamController) RegisterWith(f *DataFeed) {
+	f.RegisterForNotification(model.BlockKey{}, c.onUpdate)
+	f.RegisterForSyncStatus(c.onStatusUpdate)
+}
+
+func (c *ipamController) onStatusUpdate(s bapi.SyncStatus) {
+	c.syncStatus = s
+	switch s {
+	case bapi.InSync:
+		kick(c.syncChan)
 	}
-	log.Info("Node and IPAM data is in sync")
-	return nil
 }
 
-func (c *NodeController) gatherIPAMData() (map[string][]model.AllocationAttribute, error) {
-	log.Debug("gathering latest IPAM state")
+func (c *ipamController) onUpdate(update bapi.Update) {
+	// TODO: We need to watch nodes and populate a nodemapper
```

Comment:
Could we share the `nodemapper` from the `nodeLabelController` with the IPAM controller?

Reply:
yeah perhaps we could. It's actually the way it worked before this refactor, but was trying to get away from all of the sub-controllers being built off the same struct to clean things up a little bit. 

Maybe I can make the nodemapper its own controller which is available to both of these.

Reply:
yeah perhaps we could. It's actually the way it worked before this refactor, but was trying to get away from all of the sub-controllers being built off the same struct to clean things up a little bit. 

Maybe I can make the nodemapper its own controller which is available to both of these.